lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1623,20 +1623,37 @@ int swift::performFrontend(ArrayRef<const char *> Args,
   // arguments are generated by the driver, not directly by a user. The driver
   // is responsible for emitting diagnostics for its own errors. See SR-2683
   // for details.
-  //
-  // FIXME: Do we need one SerializedConsumer per primary? Owned by the
-  // SourceFile?
-  std::unique_ptr<DiagnosticConsumer> SerializedConsumer;
+  std::unique_ptr<DiagnosticConsumer> SerializedConsumerMultiplexer;
   {
-    const std::string &SerializedDiagnosticsPath =
-        Invocation.getSerializedDiagnosticsPathForAtMostOnePrimary();
-    if (!SerializedDiagnosticsPath.empty()) {
-      SerializedConsumer.reset(
+    SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 4>
+      SerializedConsumers;
+    Invocation.getFrontendOptions().InputsAndOutputs
+        .forEachInputProducingSupplementaryOutput(
+           [&](const InputFile &Input) -> bool {
+      std::string SerializedDiagnosticsPath = Input.serializedDiagnosticsPath();
+      if (SerializedDiagnosticsPath.empty()) {
+        // If one input does not have a serialized diagnostic output path, none
+        // of them will.
+        return true;
+      }
+
+      SerializedConsumers.emplace_back(
+          Input.file(),
           serialized_diagnostics::createConsumer(SerializedDiagnosticsPath));
-      Instance->addDiagnosticConsumer(SerializedConsumer.get());
+
+      // Continue for other inputs.
+      return false;
+    });
+
+    if (!SerializedConsumers.empty()) {
```

Comment:
It is possible to execute this block even if `SerializedConsumers` is empty? Would that just use a bit more time? It could potentially make the compiler a scintilla more reliable.

Reply:
It would waste time on every diagnostic. I'd rather not.

Reply:
It probably is negligible, but that just feels like the wrong attitude to me. The test is free and doesn't really add any complexity for a human, and it makes the implementation of FileSpecificDiagnosticConsumer have one less case to worry about. (Right now it would try to build the sorted list of consumers over and over again because I happen to be using "empty" as a sentinel state for "not yet computed".)